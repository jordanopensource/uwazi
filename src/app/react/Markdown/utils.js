const objectPath = (path, object) =>
  path.split('.').reduce((o, key) => {
    if (!o || !key) {
      return o;
    }
    return o.toJS ? o.get(key) : o[key];
  }, object);

const logError = (err, propValueOf, propLabelOf) => {
  /* eslint-disable no-console */
  console.error('Error on EntityData: ');
  console.error('value-of: ', propValueOf, '; label-of: ', propLabelOf);
  console.error(err);
  /* eslint-enable no-console */
};

const customVisualizationTags = [
  'vimeo',
  'youtube',
  'media',
  'markdownlink',
  'markdownmedia',
  'counter',
  'entitylink',
  'itemlist',
  'repeat',
  'context',
  'map',
  'value',
  'cejilchart',
  'cejilchart001',
  'cejilchart002',
  'cejilhero',
  'placeholder',
];
const customExtendedTags = [
  'searchbox',
  'contactform',
  'paypaldonatelink',
  'publicform',
  'searchbox',
  'entityinfo',
  'query',
  'dataset',
  'entitydata',
  'entitysection',
  'piechart',
  'pie',
  'piechart',
  'pie',
  'barchart',
  'bar',
  'tooltip',
  'stackeddualbarchart',
  'librarychart',
  'librarycharts',
  'coloredbar',
  'extendedtooltip',
];

const extendedValidHtmlTags = [
  'blink',
  'body',
  'button',
  'canvas',
  'embed',
  'element',
  'fieldset',
  'font',
  'foreignObject',
  'form',
  'iframe',
  'input',
  'image',
  'legend',
  'marker',
  'marquee',
  'object',
  'option',
  'select',
  'spacer',
  'style',
  'textarea',
  'title',
  'use',
  'var',
  'view',
  'wbr',
].concat(customExtendedTags);

const visualizationHtmlTags = [
  'a',
  'abbr',
  'acronym',
  'address',
  'area',
  'article',
  'aside',
  'audio',
  'b',
  'bdi',
  'bdo',
  'big',
  'blockquote',
  'br',
  'caption',
  'center',
  'circle',
  'cite',
  'clipPath',
  'code',
  'col',
  'colgroup',
  'content',
  'data',
  'datalist',
  'dd',
  'decorator',
  'defs',
  'del',
  'desc',
  'details',
  'dfn',
  'dir',
  'div',
  'dl',
  'dt',
  'ellipse',
  'em',
  'feBlend',
  'feColorMatrix',
  'feComponentTransfer',
  'feComposite',
  'feConvolveMatrix',
  'feDiffuseLighting',
  'feDisplacementMap',
  'feDistantLight',
  'feFlood',
  'feFuncA',
  'feFuncB',
  'feFuncG',
  'feFuncR',
  'feGaussianBlur',
  'feImage',
  'feMerge',
  'feMergeNode',
  'feMorphology',
  'feOffset',
  'fePointLight',
  'feSpecularLighting',
  'feSpotLight',
  'feTile',
  'feTurbulence',
  'figcaption',
  'figure',
  'filter',
  'footer',
  'g',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'head',
  'header',
  'hgroup',
  'hr',
  'html',
  'i',
  'img',
  'ins',
  'kbd',
  'label',
  'li',
  'line',
  'linearGradient',
  'main',
  'map',
  'mark',
  'mask',
  'menu',
  'menuitem',
  'metadata',
  'meter',
  'nav',
  'nobr',
  'ol',
  'optgroup',
  'output',
  'p',
  'path',
  'pattern',
  'polygon',
  'polyline',
  'pre',
  'progress',
  'q',
  'radialGradient',
  'rect',
  'rp',
  'rt',
  'ruby',
  's',
  'samp',
  'section',
  'shadow',
  'small',
  'source',
  'span',
  'strike',
  'strong',
  'sub',
  'summary',
  'sup',
  'svg',
  'switch',
  'symbol',
  'table',
  'tbody',
  'td',
  'template',
  'tfoot',
  'th',
  'thead',
  'time',
  'tr',
  'track',
  'tspan',
  'tt',
  'u',
  'ul',
  'video',
  'wbr',
].concat(customVisualizationTags);

const extendedHtmlTags = visualizationHtmlTags
  .concat(extendedValidHtmlTags)
  .concat(customExtendedTags);

export { objectPath, logError, extendedHtmlTags, visualizationHtmlTags };
