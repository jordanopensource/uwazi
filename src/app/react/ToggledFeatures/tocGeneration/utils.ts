import { ClientSettings } from 'app/apiResponseTypes';
import { SearchParams } from 'shared/types/searchParameterType';

export const tocGenerationUtils = {
  aggregations(params: SearchParams, settings: ClientSettings) {
    return {
      ...params,
      ...(settings?.features?.tocGeneration ? { aggregateGeneratedToc: true } : {}),
    };
  },
};
