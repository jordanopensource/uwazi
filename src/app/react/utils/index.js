import Jvent from 'jvent';
import rison from 'rison-node';

export function getPropsFromRoute({ routes }, componentProps) {
  const props = {};
  const lastRoute = routes[routes.length - 1];

  routes.reduceRight((_prevRoute, currRoute) => {
    componentProps.forEach(componentProp => {
      if (!props[componentProp] && currRoute.component && currRoute.component[componentProp]) {
        props[componentProp] = currRoute.component[componentProp];
      }
    });
  }, lastRoute);

  return props;
}

export function risonDecodeOrIgnore(query, defaultValue = {}) {
  try {
    return rison.decode(query);
  } catch (e) {
    console.log('Error decoding: ', query, e);
    return defaultValue;
  }
}

export const isClient = typeof document !== 'undefined';
export const events = new Jvent();
