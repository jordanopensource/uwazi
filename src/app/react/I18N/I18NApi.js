import api from 'app/utils/api';

export default {
  get(requestParams) {
    return api.get('translations', requestParams).then(response => response.json.rows);
  },

  save(requestParams) {
    return api.post('translations', requestParams).then(response => response.json);
  },

  addLanguage(requestParams) {
    return api.post('translations/languages', requestParams).then(response => response.json);
  },

  deleteLanguage(requestParams) {
    return api.delete('translations/languages', requestParams).then(response => response.json);
  },

  setDefaultLanguage(requestParams) {
    return api.post('translations/setasdeafult', requestParams).then(response => response.json);
  },

  getLanguages: async requestParams => {
    const { json: response } = await api.get('languages', requestParams);
    return response;
  },

  populateTranslations: async requestParams => {
    const { json: response } = await api.post('translations/populate', requestParams);
    return response;
  },
};
