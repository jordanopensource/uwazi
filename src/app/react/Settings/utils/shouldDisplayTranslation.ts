import { ClientLanguageSchema } from 'app/apiResponseTypes';
import { t } from 'app/I18N';
import { IImmutable } from 'shared/types/Immutable';

const shouldDisplayTranslation = (
  name: string,
  contexId: string,
  locale: string,
  availableLanguages: IImmutable<ClientLanguageSchema[]>
) => {
  const defaultLanguageKey = availableLanguages
    .find(language => language?.get('default') === true)
    .get('key');
  const isDefaultLanguage = locale === defaultLanguageKey;
  const translatedTemplateName = t(contexId, name, undefined, false);

  switch (true) {
    case isDefaultLanguage && translatedTemplateName !== name:
      return true;

    case !isDefaultLanguage:
      return true;

    default:
      return false;
  }
};

export { shouldDisplayTranslation };
