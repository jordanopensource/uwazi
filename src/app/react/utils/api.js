import { redirect } from 'react-router-dom';
import { isClient } from 'app/utils';
import { notify } from 'app/Notifications/actions/notificationsActions';
import { store } from 'app/store';
import loadingBar from 'app/App/LoadingProgressBar';
import { t } from 'app/I18N';

import { APIURL } from '../config.js';
import request from '../../shared/JSONRequest';

let API_URL = APIURL;
let language;

const doneLoading = data => {
  loadingBar.done();
  return data;
};

const isNonUsualApiError = error => error.status && ![401, 404, 409, 500].includes(error.status);

const errorMessages = [
  {
    key: /PayloadTooLargeError: request entity too large/g,
    message: 'Request data too large. Please review text length in property values',
  },
  {
    key: /ERROR Failed to index documents: (.+)/g,
    message: 'Failed to index documents: {0} ',
  },
];

function extractMessageFromError(error) {
  let finalMessage = `An error has occurred${error.json.requestId ? `, it has been logged with request id #${error.json.requestId}.` : '.'}`;
  if (!error.json.error) return finalMessage;

  const errorMessage = errorMessages.find(errorExpression =>
    error.json.error.match(errorExpression.key)
  );
  if (errorMessage) {
    const matches = errorMessage.key.exec(error.json.error);
    finalMessage = errorMessage.message;
    for (let i = 0; i < matches.length - 1; i += 1) {
      finalMessage = finalMessage.replace(`{${i}}`, matches[1]);
    }
  }
  return finalMessage;
}

function extractMessageFromValidation(error) {
  if (!error.json.validations) return error.json.error;
  return error.json.validations.reduce((message, validationError, index) => {
    const separator = index < error.json.validations.length - 1 ? ',' : '';
    return `${message} ${validationError.instancePath} ${validationError.message}${separator}`;
  }, `${error.json.error}: `);
}

// eslint-disable-next-line max-statements
const handleErrorStatus = error => {
  let errorMessage;
  switch (error.status || true) {
    case 400:
    case 422: {
      errorMessage = extractMessageFromValidation(error);
      store.dispatch(
        notify(t('System', extractMessageFromValidation(error), null, false), 'danger')
      );
      break;
    }

    case 401:
      redirect('/login');
      break;

    case 404:
      redirect('/404');
      break;

    case 409: {
      errorMessage = error.json.error;
      store.dispatch(notify(t('System', error.json.error, null, false), 'warning'));
      break;
    }

    case 500: {
      errorMessage = extractMessageFromError(error);
      store.dispatch(notify(t('System', extractMessageFromError(error), null, false), 'danger'));
      break;
    }

    case isNonUsualApiError(error): {
      errorMessage = error.json.prettyMessage || error.json.error;
      store.dispatch(
        notify(t('System', error.json.prettyMessage || error.json.error, null, false), 'danger')
      );
      break;
    }

    case error instanceof TypeError: {
      errorMessage = 'Could not reach server. Please try again later.';
      store.dispatch(
        notify(
          t('System', 'Could not reach server. Please try again later.', null, false),
          'danger'
        )
      );
      break;
    }

    default: {
      errorMessage = 'An error occurred';
      store.dispatch(notify(t('System', 'An error occurred', null, false), 'danger'));
    }
  }

  return errorMessage;
};

const handleError = (e, endpoint) => {
  const error = e;
  error.endpoint = endpoint;

  if (!isClient) {
    return Promise.reject(error);
  }

  doneLoading();

  handleErrorStatus(error);
  return Promise.reject(error);
};

const _request = (url, req, method) => {
  const headers = {
    'Content-Language': language,
    ...req.headers,
    'X-Requested-With': 'XMLHttpRequest',
  };

  loadingBar.start();

  return request[method](API_URL + url, req.data, headers)
    .then(doneLoading)
    .catch(async e => {
      await handleError(e, { url, method });
    });
};

export default {
  get: (url, req = {}) => _request(url, req, 'get'),

  post: (url, req = {}) => _request(url, req, 'post'),

  put: (url, req = {}) => _request(url, req, 'put'),

  delete: (url, req = {}) => _request(url, req, 'delete'),

  cookie(c) {
    request.cookie(c);
  },

  locale(locale) {
    language = locale;
  },

  APIURL(url) {
    API_URL = url;
  },
};
