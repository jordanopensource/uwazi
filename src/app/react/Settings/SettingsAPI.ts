import api from 'app/utils/api';
import { RequestParams } from 'app/utils/RequestParams';
import { ClientSettings } from 'app/apiResponseTypes';

export default {
  async save(settings: RequestParams<ClientSettings>): Promise<ClientSettings> {
    return api.post('settings', settings).then((response: any) => response.json);
  },

  async get(query: RequestParams = new RequestParams()): Promise<ClientSettings> {
    return api.get('settings', query).then((response: any) => response.json);
  },

  async stats(query: RequestParams = new RequestParams()) {
    return api.get('stats', query).then((response: any) => response.json);
  },
};
