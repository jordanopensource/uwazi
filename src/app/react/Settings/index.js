/* eslint-disable import/no-named-as-default */
import EntityTypesList from './components/EntityTypesList';
import { Settings } from './Settings';
import SettingsAPI from './SettingsAPI';
import { PreserveSettings } from './components/PreserveSettings';

export { Settings, SettingsAPI, PreserveSettings, EntityTypesList };
