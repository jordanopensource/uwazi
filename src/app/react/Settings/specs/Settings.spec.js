import React from 'react';
import { shallow } from 'enzyme';
import UsersAPI from 'app/Users/UsersAPI';
import ThesauriAPI from 'app/Thesauri/ThesauriAPI';
import { I18NApi } from 'app/I18N';
import { RequestParams } from 'app/utils/RequestParams';
import TemplatesAPI from 'app/Templates/TemplatesAPI';
import api from 'app/Search/SearchAPI';
import SettingsAPI from '../SettingsAPI';
import { SettingsComponent as Settings } from '../Settings';

describe('Settings', () => {
  let user;
  const dictionaries = [{ _id: 1, name: 'Countries' }];
  const models = [];
  const translations = [{ _id: 1, locale: 'es', values: {} }];
  const templates = [];
  const settings = { siteName: 'BatCave' };
  const suggestions = {};

  beforeEach(() => {
    user = { name: 'doe', role: 'admin' };
    spyOn(UsersAPI, 'currentUser').and.callFake(async () => Promise.resolve(user));
    spyOn(ThesauriAPI, 'getThesauri').and.callFake(async () => Promise.resolve(dictionaries));
    spyOn(ThesauriAPI, 'getModelStatus').and.callFake(async () => Promise.resolve(models));
    spyOn(I18NApi, 'get').and.callFake(async () => Promise.resolve(translations));
    spyOn(SettingsAPI, 'get').and.callFake(async () => Promise.resolve(settings));
    spyOn(SettingsAPI, 'stats').and.callFake(async () => Promise.resolve({ files: 3, users: 2 }));
    spyOn(TemplatesAPI, 'get').and.callFake(async () => Promise.resolve(templates));
    spyOn(api, 'search').and.callFake(async () => Promise.resolve(suggestions));
  });

  describe('requestState', () => {
    it('should get the current user, and metadata', async () => {
      const actions = await Settings.requestState(new RequestParams());
      expect(actions).toMatchSnapshot();
    });

    it('should not request the stats if the user does not have admin role', async () => {
      user = { name: 'jane', role: 'editor' };
      const actions = await Settings.requestState(new RequestParams());
      expect(actions).toMatchSnapshot();
    });
  });

  describe('render', () => {
    let component;
    const render = children => {
      const context = { store: { getState: () => ({}), dispatch: jasmine.createSpy('dispatch') } };
      component = shallow(<Settings>{children}</Settings>, { context });
    };

    it('should add the only-desktop class when recieving children only', () => {
      render('Account settings');
      expect(component).toMatchSnapshot();

      render();
      expect(component).toMatchSnapshot();
    });
  });
});
