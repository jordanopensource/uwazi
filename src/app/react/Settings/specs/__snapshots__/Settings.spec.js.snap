// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings render should add the only-desktop class when recieving children only 1`] = `
<div
  className="row settings"
>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Settings
    </title>
  </HelmetWrapper>
  <div
    className="settings-navigation "
  >
    <Connect(SettingsNavigationComponent) />
  </div>
  <Outlet />
</div>
`;

exports[`Settings render should add the only-desktop class when recieving children only 2`] = `
<div
  className="row settings"
>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Settings
    </title>
  </HelmetWrapper>
  <div
    className="settings-navigation "
  >
    <Connect(SettingsNavigationComponent) />
  </div>
  <Outlet />
</div>
`;

exports[`Settings requestState should get the current user, and metadata 1`] = `
Array [
  Object {
    "type": "auth/user/SET",
    "value": Object {
      "name": "doe",
      "role": "admin",
    },
  },
  Object {
    "type": "dictionaries/SET",
    "value": Array [
      Object {
        "_id": 1,
        "name": "Countries",
      },
    ],
  },
  Object {
    "type": "translations/SET",
    "value": Array [
      Object {
        "_id": 1,
        "locale": "es",
        "values": Object {},
      },
    ],
  },
  Object {
    "type": "settings/collection/SET",
    "value": Object {
      "siteName": "BatCave",
    },
  },
  Object {
    "type": "settings/stats/SET",
    "value": Object {
      "files": 3,
      "users": 2,
    },
  },
]
`;

exports[`Settings requestState should not request the stats if the user does not have admin role 1`] = `
Array [
  Object {
    "type": "auth/user/SET",
    "value": Object {
      "name": "jane",
      "role": "editor",
    },
  },
  Object {
    "type": "dictionaries/SET",
    "value": Array [
      Object {
        "_id": 1,
        "name": "Countries",
      },
    ],
  },
  Object {
    "type": "translations/SET",
    "value": Array [
      Object {
        "_id": 1,
        "locale": "es",
        "values": Object {},
      },
    ],
  },
  Object {
    "type": "settings/collection/SET",
    "value": Object {
      "siteName": "BatCave",
    },
  },
  Object {
    "type": "settings/stats/SET",
    "value": Object {},
  },
]
`;
